drivers/src/pci/capabilities.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use arrayvec::ArrayVec;
+use core::ptr::NonNull;
+use tock_registers::interfaces::Readable;
+
+use super::registers::*;
+
+// Standard PCI capability IDs.
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum CapabilityId {
+    PowerManagement = 1,
+    Msi = 5,
+    Vendor = 9,
+    BridgeSubsystem = 13,
+    PciExpress = 16,
+    MsiX = 17,
+}
+
+impl CapabilityId {
+    // Returns the `CapabilityId` from the raw register value.
+    fn from_raw(id: u8) -> Option<Self> {
+        use CapabilityId::*;
+        match id {
+            1 => Some(PowerManagement),
+            5 => Some(Msi),
+            9 => Some(Vendor),
+            13 => Some(BridgeSubsystem),
+            16 => Some(PciExpress),
+            17 => Some(MsiX),
+            _ => None,
+        }
+    }
+}
+
+// Maps a PCI capability ID to its offset in config space.
+#[derive(Clone, Copy, Debug)]
+struct PciCapability {
+    id: CapabilityId,
+    offset: usize,
+}
+
+// The maximum number of capabilities we support for a single device. Enough for the typical PCI
+// devices virtualized by QEMU.
+const MAX_PCI_CAPS: usize = 8;
+
+/// Maps the location of PCI capabilities in a device's config space and handles emulation of
+/// reads and writes to these capabilities.
+pub struct PciCapabilities {
+    caps: ArrayVec<PciCapability, MAX_PCI_CAPS>,
+}
+
+impl PciCapabilities {
+    /// Creates a new `PciCapabilities` by parsing the PCI capability linked-list starting at
+    /// `start_offset` within the standard PCI configuration space pointed to be `config_ptr`.
+    ///
+    /// # Safety
+    ///
+    /// `config_ptr` must refer to a valid and uniquely-owned PCI configuration space.
```

Comment:
what does the caller check to make sure that CommonRegisters are valid? Could that be encapsulated in the type (meaning only valid CommonRegisters can exist)? If so, can this function take a reference to a CommonRegisters and be safe knowing that ref is valid?